Prevent full rebuilds when rerunning `flutter test`.

victoreronmosele opened the pull request:
This PR updates `FlutterPlatform` to prevent a rebuild of the assemble targets when rerunning 'flutter test'.

This update changes the directory where the `listener.dart` file is created from the system temp folder to a location based on the test file hash: `build/{fileHash}` folder.  

Since the location of the `listener.dart` file is now [stable](https://github.com/flutter/flutter/issues/108400#issuecomment-1196104701), the build system no longer creates a different temp directory on every run.  

This means the file store hash for the `listener.dart` file [stays the same](https://github.com/victoreronmosele/flutter/blob/bec1d139454d871b5bdd3eba8f8901ca3eee47b2/packages/flutter_tools/lib/src/build_system/build_system.dart#L1090-L1101), and rebuilds are [skipped](https://github.com/victoreronmosele/flutter/blob/bec1d139454d871b5bdd3eba8f8901ca3eee47b2/packages/flutter_tools/lib/src/build_system/build_system.dart#L839).

Fixes #108400 

## Test Results

Rerunning this command for an integration test on an iOS simulator 

```dart
flutter --no-color test --machine  --plain-name testGetVisibleRegion  -v integration_test/main_test.dart
```                                 
gave me these results below:

- _After Update:_ 

  <details>
  <summary>flutter assemble" took 1,114ms. Skipped all targets.</summary>

  #### Details:

  ```bash
  /Users/victoreronmosele/development/mobile_projects/flutter_projects/flutter_contributions/flutter/bin/flutter --verbose assemble
  --no-version-check
  --output=/Users/victoreronmosele/development/mobile_projects/flutter_projects/demo/build/ios/Debug-iphonesimulator/
  -dTargetPlatform=ios
  -dTargetFile=/Users/victoreronmosele/development/mobile_projects/flutter_projects/demo/build/flutter_test_listener/listener.dart
  -dBuildMode=debug -dIosArchs=arm64
  -dSdkRoot=/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator16.4.sdk
  -dSplitDebugInfo= -dTreeShakeIcons=false -dTrackWidgetCreation=true -dDartObfuscation=false -dAction=build
  -dFrontendServerStarterPath= --ExtraGenSnapshotOptions=
  --DartDefines=RkxVVFRFUl9XRUJfQVVUT19ERVRFQ1Q9dHJ1ZQ==,SU5URUdSQVRJT05fVEVTVF9TSE9VTERfUkVQT1JUX1JFU1VMVFNfVE9fTkFUSVZFPWZhbHNl
  --ExtraFrontEndOptions= -dCodesignIdentity=- debug_ios_bundle_flutter_assets
  Resolving dependencies...
  Got dependencies.
  [   +1 ms] Could not interpret results of "git describe": 3.18.0-2.0.pre.1
  [  +16 ms] executing: sysctl hw.optional.arm64
  [   +3 ms] Exit code 0 from: sysctl hw.optional.arm64
  [        ] hw.optional.arm64: 1
  [  +18 ms] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
  [        ] Artifact Instance of 'LegacyCanvasKitRemover' is not required, skipping update.
  [   +1 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
  [  +47 ms] Artifact Instance of 'MaterialFonts' is not required, skipping update.
  [        ] Artifact Instance of 'GradleWrapper' is not required, skipping update.
  [        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
  [        ] Artifact Instance of 'LegacyCanvasKitRemover' is not required, skipping update.
  [        ] Artifact Instance of 'FlutterSdk' is not required, skipping update.
  [        ] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
  [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'FontSubsetArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'PubDependencies' is not required, skipping update.
  [  +10 ms] Initializing file store
  [   +5 ms] Done initializing file store
  [ +289 ms] Skipping target: debug_unpack_ios
  [   +2 ms] Skipping target: native_assets
  [        ] Skipping target: gen_localizations
  [        ] Skipping target: gen_dart_plugin_registrant
  [ +439 ms] Skipping target: kernel_snapshot
  [   +2 ms] Skipping target: debug_universal_framework
  [ +295 ms] Skipping target: debug_ios_bundle_flutter_assets
  [        ] Persisting file store
  [   +5 ms] Done persisting file store
  [   +4 ms] build succeeded.
  [   +3 ms] "flutter assemble" took 1,114ms.
  ```

  </details>



- _Before Update:_ 

  <details>
  <summary>flutter assemble" took 5,583ms. Didn't skip all targets.</summary>

  #### Details:

  ```bash
  /Users/victoreronmosele/development/mobile_projects/flutter_projects/flutter_contributions/flutter/bin/flutter --verbose assemble
  --no-version-check
  --output=/Users/victoreronmosele/development/mobile_projects/flutter_projects/demo/build/ios/Debug-iphonesimulator/
  -dTargetPlatform=ios
  -dTargetFile=/var/folders/8n/qpsf1kw9371gl1_vvgt4_05r0000gn/T/flutter_tools.NjiIAc/flutter_test_listener.cQySHc/listener.dart
  -dBuildMode=debug -dIosArchs=arm64
  -dSdkRoot=/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator16.4.sdk
  -dSplitDebugInfo= -dTreeShakeIcons=false -dTrackWidgetCreation=true -dDartObfuscation=false -dAction=build
  -dFrontendServerStarterPath= --ExtraGenSnapshotOptions=
  --DartDefines=RkxVVFRFUl9XRUJfQVVUT19ERVRFQ1Q9dHJ1ZQ==,SU5URUdSQVRJT05fVEVTVF9TSE9VTERfUkVQT1JUX1JFU1VMVFNfVE9fTkFUSVZFPWZhbHNl
  --ExtraFrontEndOptions= -dCodesignIdentity=- debug_ios_bundle_flutter_assets
  [   +2 ms] Could not interpret results of "git describe": 3.18.0-2.0.pre.1
  [  +11 ms] executing: sysctl hw.optional.arm64
  [   +3 ms] Exit code 0 from: sysctl hw.optional.arm64
  [        ] hw.optional.arm64: 1
  [  +10 ms] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
  [        ] Artifact Instance of 'LegacyCanvasKitRemover' is not required, skipping update.
  [        ] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
  [  +45 ms] Artifact Instance of 'MaterialFonts' is not required, skipping update.
  [        ] Artifact Instance of 'GradleWrapper' is not required, skipping update.
  [        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
  [        ] Artifact Instance of 'LegacyCanvasKitRemover' is not required, skipping update.
  [        ] Artifact Instance of 'FlutterSdk' is not required, skipping update.
  [        ] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
  [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'FontSubsetArtifacts' is not required, skipping update.
  [        ] Artifact Instance of 'PubDependencies' is not required, skipping update.
  [  +11 ms] Initializing file store
  [  +12 ms] debug_unpack_ios: Starting due to {InvalidatedReasonKind.inputChanged: The following inputs have updated contents:
  /Users/victoreronmosele/development/mobile_projects/flutter_projects/flutter_contributions/flutter/packages/flutter_tools/lib/src/bu
  ild_system/targets/ios.dart,/Users/victoreronmosele/development/mobile_projects/flutter_projects/flutter_contributions/flutter/bin/i
  nternal/engine.version}
  [   +3 ms] native_assets: Starting due to {}
  [        ] Skipping target: gen_localizations
  [   +1 ms] gen_dart_plugin_registrant: Starting due to {InvalidatedReasonKind.inputChanged: The following inputs have updated
  contents: /Users/victoreronmosele/development/mobile_projects/flutter_projects/demo/.dart_tool/package_config_subset}
  Directory: '/Users/victoreronmosele/development/mobile_projects/flutter_projects/demo'
  does file exist? true
  uri is file:///Users/victoreronmosele/development/mobile_projects/flutter_projects/demo/.dart_tool/package_config.json
  uri path is /Users/victoreronmosele/development/mobile_projects/flutter_projects/demo/.dart_tool/package_config.json
  uri path segments are [Users, victoreronmosele, development, mobile_projects, flutter_projects, demo, .dart_tool,
  package_config.json]
  uri path without using fromUri exists? true
  uri path completed wtih current path directory exist? false
  [  +32 ms] Writing native_assets.yaml.
  [   +7 ms] Writing
  /Users/victoreronmosele/development/mobile_projects/flutter_projects/demo/.dart_tool/flutter_build/92e76f94a27f20a6898bdaebc3fbe32b/
  native_assets.yaml done.
  [   +1 ms] native_assets: Complete
  [  +25 ms] Found plugin integration_test at
  /Users/victoreronmosele/development/mobile_projects/flutter_projects/flutter_contributions/flutter/packages/integration_test/
  [   +4 ms] gen_dart_plugin_registrant: Complete
  [        ] kernel_snapshot: Starting due to {}
  [        ] Embedding native assets mapping
  /Users/victoreronmosele/development/mobile_projects/flutter_projects/demo/.dart_tool/flutter_build/92e76f94a27f20a6898bdaebc3fbe32b/
  native_assets.yaml in kernel.
  [   +2 ms]
  /Users/victoreronmosele/development/mobile_projects/flutter_projects/flutter_contributions/flutter/bin/cache/dart-sdk/bin/dartaotrun
  time --disable-dart-dev
  /Users/victoreronmosele/development/mobile_projects/flutter_projects/flutter_contributions/flutter/bin/cache/dart-sdk/bin/snapshots/
  frontend_server_aot.dart.snapshot --sdk-root
  /Users/victoreronmosele/development/mobile_projects/flutter_projects/flutter_contributions/flutter/bin/cache/artifacts/engine/common
  /flutter_patched_sdk/ --target=flutter --no-print-incremental-dependencies -DFLUTTER_WEB_AUTO_DETECT=true
  -DINTEGRATION_TEST_SHOULD_REPORT_RESULTS_TO_NATIVE=false -Ddart.vm.profile=false -Ddart.vm.product=false --enable-asserts
  --track-widget-creation --no-link-platform --packages
  /Users/victoreronmosele/development/mobile_projects/flutter_projects/demo/.dart_tool/package_config.json --output-dill
  /Users/victoreronmosele/development/mobile_projects/flutter_projects/demo/.dart_tool/flutter_build/92e76f94a27f20a6898bdaebc3fbe32b/
  app.dill --depfile
  /Users/victoreronmosele/development/mobile_projects/flutter_projects/demo/.dart_tool/flutter_build/92e76f94a27f20a6898bdaebc3fbe32b/
  kernel_snapshot.d --incremental --initialize-from-dill
  /Users/victoreronmosele/development/mobile_projects/flutter_projects/demo/.dart_tool/flutter_build/92e76f94a27f20a6898bdaebc3fbe32b/
  app.dill --native-assets
  /Users/victoreronmosele/development/mobile_projects/flutter_projects/demo/.dart_tool/flutter_build/92e76f94a27f20a6898bdaebc3fbe32b/
  native_assets.yaml --verbosity=error
  file:///var/folders/8n/qpsf1kw9371gl1_vvgt4_05r0000gn/T/flutter_tools.NjiIAc/flutter_test_listener.cQySHc/listener.dart
  [ +879 ms] executing: xattr -r -d com.apple.FinderInfo
  /Users/victoreronmosele/development/mobile_projects/flutter_projects/demo/build/ios/Debug-iphonesimulator/Flutter.framework/Flutter
  [ +230 ms] debug_unpack_ios: Complete
  [+2674 ms] kernel_snapshot: Complete
  [ +369 ms] debug_universal_framework: Starting due to {InvalidatedReasonKind.inputChanged: The following inputs have updated
  contents:
  /Users/victoreronmosele/development/mobile_projects/flutter_projects/flutter_contributions/flutter/packages/flutter_tools/lib/src/bu
  ild_system/targets/ios.dart}
  [   +7 ms] executing: sysctl hw.optional.arm64
  [   +4 ms] Exit code 0 from: sysctl hw.optional.arm64
  [        ] hw.optional.arm64: 1
  [        ] executing: /usr/bin/arch -arm64e xcrun clang -x c -arch arm64
  /var/folders/8n/qpsf1kw9371gl1_vvgt4_05r0000gn/T/flutter_tools.91jToJ/flutter_tools_stub_source.uWBMT9/debug_app.cc -dynamiclib
  -miphonesimulator-version-min=11.0 -Xlinker -rpath -Xlinker @executable_path/Frameworks -Xlinker -rpath -Xlinker
  @loader_path/Frameworks -fapplication-extension -install_name @rpath/App.framework/App -isysroot
  /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator16.4.sdk -o
  /Users/victoreronmosele/development/mobile_projects/flutter_projects/demo/.dart_tool/flutter_build/92e76f94a27f20a6898bdaebc3fbe32b/
  App.framework/App
  [ +108 ms] executing: xattr -r -d com.apple.FinderInfo
  /Users/victoreronmosele/development/mobile_projects/flutter_projects/demo/.dart_tool/flutter_build/92e76f94a27f20a6898bdaebc3fbe32b/
  App.framework/App
  [  +15 ms] debug_universal_framework: Complete
  [   +2 ms] debug_ios_bundle_flutter_assets: Starting due to {}
  [ +153 ms] shaderc command:
  [/Users/victoreronmosele/development/mobile_projects/flutter_projects/flutter_contributions/flutter/bin/cache/artifacts/engine/darwi
  n-x64/impellerc, --runtime-stage-metal, --iplr,
  --sl=/Users/victoreronmosele/development/mobile_projects/flutter_projects/demo/build/ios/Debug-iphonesimulator/App.framework/flutter
  _assets/shaders/ink_sparkle.frag,
  --spirv=/Users/victoreronmosele/development/mobile_projects/flutter_projects/demo/build/ios/Debug-iphonesimulator/App.framework/flut
  ter_assets/shaders/ink_sparkle.frag.spirv,
  --input=/Users/victoreronmosele/development/mobile_projects/flutter_projects/flutter_contributions/flutter/packages/flutter/lib/src/
  material/shaders/ink_sparkle.frag, --input-type=frag,
  --include=/Users/victoreronmosele/development/mobile_projects/flutter_projects/flutter_contributions/flutter/packages/flutter/lib/sr
  c/material/shaders,
  --include=/Users/victoreronmosele/development/mobile_projects/flutter_projects/flutter_contributions/flutter/bin/cache/artifacts/eng
  ine/darwin-x64/shader_lib]
  [ +159 ms] executing: xattr -r -d com.apple.FinderInfo
  /Users/victoreronmosele/development/mobile_projects/flutter_projects/demo/build/ios/Debug-iphonesimulator/App.framework/App
  [ +315 ms] debug_ios_bundle_flutter_assets: Complete
  [ +491 ms] Persisting file store
  [   +6 ms] Done persisting file store
  [   +5 ms] build succeeded.
  [   +5 ms] "flutter assemble" took 5,583ms.
  ```
  </details>

<details>
<summary>Integration Test</summary>

#### Details:

```dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:integration_test/integration_test.dart';
import 'package:demo/main.dart' as app;

void main() {
  IntegrationTestWidgetsFlutterBinding.ensureInitialized();

  testWidgets('Verify counter increments when FAB is pressed', (WidgetTester tester) async {
    app.main();

    await tester.pumpAndSettle();

    expect(find.text('0'), findsOneWidget);

    await tester.tap(find.byType(FloatingActionButton));
    await tester.pumpAndSettle();

    expect(find.text('1'), findsOneWidget);
  });
}
```

</details>

<details>
<summary>App's Entrypoint (main.dart)</summary>

#### Details:

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headlineMedium,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), 
    );
  }
}
```

</details>



## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

victoreronmosele committed:
Update listener.dart location to project's build folder instead of temp folder.

DanTup commented:
> Since the location of the listener.dart file is now https://github.com/flutter/flutter/issues/108400#issuecomment-1196104701, the build system no longer creates a different temp directory on every run.

Would this have any impact on concurrent runs of `flutter test`? (in VS Code, there are a few ways you could trigger tests to be run in multiple processes at the same time).

christopherfujino commented:
> > Since the location of the listener.dart file is now [#108400 (comment)](https://github.com/flutter/flutter/issues/108400#issuecomment-1196104701), the build system no longer creates a different temp directory on every run.
> 
> Would this have any impact on concurrent runs of `flutter test`? (in VS Code, there are a few ways you could trigger tests to be run in multiple processes at the same time).

@victoreronmosele this seems like a very valid question.

victoreronmosele commented:
@DanTup @christopherfujino just got some time to look into this. 

I'll share what I find out shortly.

victoreronmosele commented:
@DanTup @christopherfujino 

After testing, it turns out this change affects concurrent tests and sequential tests when you are testing multiple test files. 

It sometimes crashes with this error: 

```
Failed to load "path/to/a_test.dart": Shell subprocess crashed with unexpected exit code -10.
```

I believe this is because the `listener.dart` file is meant to be created for each test file but in the current state, it is being reused across different tests. 

To fix this, I'm making a change to store each test file's `listener.dart` in a path unique to the test file (in a directory with the file hash as the name). 

So the `listener.dart` path is changed from: 

`build/flutter_test_listener/listener.dart`

to: 

`build/{fileHash}/listener.dart`.

I've tested this fix and the crash no longer happens. 

I'll update the PR after I've updated the test cases.

victoreronmosele committed:
Create test listener in unique path for each test file

victoreronmosele commented:
I'm currently looking into the failing tests.

victoreronmosele commented:
Concurrent and sequential tests run across single and multiple test files will no longer be affected as each test file now has it's own listener.dart file based on its hash. 

@DanTup @christopherfujino

DanTup commented:
@victoreronmosele looks like there are currently some red bots.

The change seems reasonable to me, though I'm not super familiar with this listener file or possible implications of moving the file into the project, so I'll defer to @christopherfujino for approving.

christopherfujino commented:
@dnfield is this:

1. safe to cache?
2. compatible with our multi-engine refactor coming down the line?

rick-alp reviewed (commented):
Hello, Thanks for the work on this!  What is the status? Is this moving forward?

StadesaSystems commented:
All flutter community i relying on you. 🙏🙏🙏

dnfield commented:
> @dnfield is this:
> 
> 1. safe to cache?

Yes, I believe it is.

> 2. compatible with our multi-engine refactor coming down the line?

That uses a different code path entirely, which coincidentally puts the generated test script under the build directory as well (/cc @derekxu16 )

dnfield commented:
The test expectations need to be updated if we want to land this. Right now `packages/flutter_tools/test/integration.shard/test_test.dart` has a test that asserts this command logs about deleting a temp directory, and that fails on this patch because there's no temp directory to delete.

victoreronmosele commented:
> The test expectations need to be updated if we want to land this. Right now `packages/flutter_tools/test/integration.shard/test_test.dart` has a test that asserts this command logs about deleting a temp directory, and that fails on this patch because there's no temp directory to delete.

I'll update the test expectations. Thanks for identifying the cause of failure @dnfield!

victoreronmosele committed:
Update flutter test directory deletion log message in test expectations

victoreronmosele commented:
I've updated the test expectations, and the previously failing tests concerning deleting the test listener file's directory are now passing. 

Now, only the Windows tool integration test is failing at the test: [flutter test should serve DevTools](https://github.com/victoreronmosele/flutter/blob/79fb69a536d85728f9c886bf7a346485330097e8/packages/flutter_tools/test/integration.shard/test_test.dart#L420-L421). 

I can't see a direct correlation between the failing test and the current changes. 🤔

victoreronmosele commented:
The [flutter test should serve DevTools](https://github.com/victoreronmosele/flutter/blob/79fb69a536d85728f9c886bf7a346485330097e8/packages/flutter_tools/test/integration.shard/test_test.dart#L420-L421) test passes when run in isolation but it times out when run with the other tests in the file.

victoreronmosele commented:
> The [flutter test should serve DevTools](https://github.com/victoreronmosele/flutter/blob/79fb69a536d85728f9c886bf7a346485330097e8/packages/flutter_tools/test/integration.shard/test_test.dart#L420-L421) test on Windows passes when run in isolation but it times out when run with the other tests in the file.

This test was failing due to this line in the `TestCompiler`.

https://github.com/victoreronmosele/flutter/blob/70f0845bb60e2f1ee9c909b7c7ce66c57b4b6d14/packages/flutter_tools/lib/src/test/test_compiler.dart#L202

An Exception is thrown when trying to copy the content of `outputFile` to `$path.dill` when the `$path.dill` file already exists.

My solution is to delete the existing `$path.dill` file before copying the `outputFile` to that location.

victoreronmosele committed:
Reuse test listener dill file if it already exists

christopherfujino commented:
@andrewkolos can you review this?

andrewkolos commented on packages/flutter_tools/lib/src/test/test_compiler.dart:
```diff
@@ -194,7 +194,15 @@ class TestCompiler {
           final File outputFile = globals.fs.file(outputPath);
-          final File kernelReadyToRun = await outputFile.copy('$path.dill');
+
+          final File kernelReadyToRun;
+
+          if (globals.fs.file('$path.dill').existsSync()) {
+            kernelReadyToRun = globals.fs.file('$path.dill');
+          } else {
+            kernelReadyToRun = await outputFile.copy('$path.dill');
+          }
```
Should we instead overwrite the already existing file? I am not a subject matter expert when it comes to `flutter test`, so I am afraid of a scenario (now or in the future) where the `.dill` file already exists but has outdated contents. I think that if we instead change this to first delete the old file if it exists, we protect ourselves from such a situation at a fairly small performance cost (I'm assuming these `.dill` files aren't massive in size, but I am not certain).

DanTup replied:
> if we instead change this to first delete the old file if it exists

I don't know enough about this code to know if this is a valid concern, but if there are multiple parallel invocations of running tests that might use the same file, deleting and then re-creating might impact other runs if they were reading the file?

(comments above suggested using a hash in the filename.. I don't know if that applies to this `$path` here, but if so perhaps that avoids the issue described here?)

andrewkolos replied:
> (comments above suggested using a hash in the filename.. I don't know if that applies to this `$path` here, but if so perhaps that avoids the issue described here?)

Ah yes, this slipped my mind. I think this should be fine as-is then.

andrewkolos replied:
Ah wait, I remember my original thoughts now. Something I need to investigate is whether or not the `.dill` file will always have a hash (even outside of scenarios where we use `createListenerDart`).

victoreronmosele replied:
The listener file currently uses a hash based on the test file path so that the location is stable and it can be reused when rerunning tests. 

> I am afraid of a scenario (now or in the future) where the .dill file already exists but has outdated contents.

What if we make the hash based on both the test file path and the content (of the `_generateTestMain()`), that way we can't have outdated contents since the path will be unique to the file's content?

victoreronmosele replied:
> (comments above suggested using a hash in the filename.. I don't know if that applies to this $path here, but if so perhaps that avoids the issue described here?)

yes, the hash applies to `$path`. 

Here's how the hash is used in the build folder:
- `build/6746d8422dfd3180d50b95edd160f5f8/listener.dart`
- `build/6746d8422dfd3180d50b95edd160f5f8/listener.dart.dill`

andrewkolos replied:
To summarize my thoughts since it's been a while, writing the dill file only if it doesn't already exist creates a pitfall that could result in the file not being overwritten despite it being out-of-date. As a hypothetical scenario, consider some new code instantiates a `TestCompiler` and calls `compile` on it twice with the same file path (but with different file contents), the dill will be from the first compile operation, not the second. I haven't confirmed this, but from reading the code I believe this is the case.

> What if we make the hash based on both the test file path and the content (of the `_generateTestMain()`), that way we can't have outdated contents since the path will be unique to the file's content?

I don't think hashing on the content of just the one Dart file would do much here as the file could import other files that can change. However, I don't think this matters much, because any situation where we overwrite the file is susceptible to whatever [issues you ran into earlier with the `flutter test should serve DevTools`](https://github.com/flutter/flutter/pull/140284/#issuecomment-1977014359) test, no? 

I imagine a more robust solution to https://github.com/flutter/flutter/issues/108400 would be to change [the logic the build system uses to invalidate builds when defines change](https://github.com/flutter/flutter/blob/61a0add2865c51bfee33939c1820709d1115c77d/packages/flutter_tools/lib/src/build_system/build_system.dart#L351-L358). However, this would be a lot more work.

Perhaps one simpler solution would be have the `TestCompiler::compile` method accept an output destination as a parameter. Then, `FlutterPlatform` can provide a random and temporary filename. Would this or some variant of this work?

victoreronmosele replied:
I think this is a viable solution since the `listener. dart` file location is the one we need stable to prevent rebuilds and not the kernel file location.

I'll implement this.

victoreronmosele replied:
I implemented a solution in 89c1285 using `globals.fsUtils.getUniqueFile` to make the dill file unique.

andrewkolos commented on packages/flutter_tools/test/general.shard/flutter_platform_test.dart:
```diff
@@ -121,5 +122,134 @@ void main() {
+          'test.dart',
+        );
+
+        const String expectedListenerFilePath =
+            '/build/068070b05f23393259df743349b63e04/listener.dart';
```
Minor style nit: [prefer avoiding line breaks after assignment operators](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#prefer-a-maximum-line-length-of-80-characters)

```suggestion
        const String expectedListenerFilePath = '/build/068070b05f23393259df743349b63e04/listener.dart';
```

victoreronmosele committed:
Refactor expectedListenerFilePath assignment to match style guide

andrewkolos commented:
FYI I updated this branch to include the fix from https://github.com/flutter/flutter/commit/6967ae551eaff91e2d0b483f31564993803282d6.

andrewkolos commented:
Thanks for contributing this. Since this PR has gone several weeks without a response to feedback, I'll be closing this PR just to get it off of our PR review queue. Feel free to reopen or  submit a new PR if you have the time to address the review comments.

victoreronmosele commented:
Sorry for the long break on this PR, and thanks for the update, @andrewkolos. 

I'll address the feedback very soon.

victoreronmosele committed:
Store test kernel file in a unique location.

This is to prevent an error (on Windows) when copying the compiler output to the test kernel file path.

See https://github.com/flutter/flutter/pull/140284/#issuecomment-1977014359

victoreronmosele commented:
I've addressed the feedback in https://github.com/flutter/flutter/pull/140284/#discussion_r1677435045.

Could you reopen the PR? I don't have permission to do that on this repository.

victoreronmosele committed:
Add required packageConfigPath to BuildInfo in flutter_platform_test.dart

andrewkolos commented:
Thanks for all of your effort in this contribution. When taking into account the 1) current top priorities of the Flutter CLI (quality/reliability), 2) the complexity/riskiness of this change, and 3) the current bandwidth of the Flutter tools team, I regrettably believe the best thing to do is to close this PR and focus on PRs that address higher priority issues (or that otherwise have a better complexity-vs-value ratio). I cannot make any promises as to when myself or another contributor would have the bandwidth to revisit the linked issue.

I am deeply sorry to have strung this along for so long.

victoreronmosele commented:
Thanks for the explanation, @andrewkolos! 

I completely understand the situation.
